// indexer/indexer.js
var esprima = require('esprima');
var esrefactor = require('esrefactor');
var estraverse = require('estraverse');

function indexFile(code, file, verbose)
{
    if (code.lastIndexOf("#!", 0) === 0) {
        var ch = 0;
        var header = "";
        while (ch < code.length && code.charCodeAt(ch) !== 10) {
            ++ch;
            header += " ";
        }
        code = header + code.substring(ch);
        console.log("replaced", ch);
    }
    var esrefactorContext = new esrefactor.Context();
    var parsed;
    try {
        parsed = esprima.parse(code, { tolerant: true, range: true });
    } catch (err) {
        console.log("Got error", err, "for", code);
        return {errors:[err]};
    }

    if (!parsed) {
        throw new Error("Couldn't parse file " + file + ' ' + code.length);
        return undefined;
    }
    esrefactorContext.setCode(parsed);
    if (!esrefactorContext._syntax)
        throw new Error('Unable to identify anything without a syntax tree');

    var parents = [];

    // console.log("BALLS", JSON.stringify(parsed, null, 4));

    function location(range)
    {
        return code.substring(range[0], range[1]);
    }
    function childKey(child) // slow
    {
        var p = child.parent;
        for (var key in p) {
            if (p[key] === child) {
                return key;
            }
        }
        return "not found";
    }
    function isChild(key, node)
    {
        if (!node)
            node = parents[parents.length - 1];

        if (node.parent) {
            var p = node.parent[key];
            if (p !== undefined) {
                if (Array.isArray(p)) {
                    return p.indexOf(node) != -1;
                } else {
                    return p == node;
                }
            }
        }
        return false;
    }
    function parentTypeIs(type, node)
    {
        if (!node)
            node = parents[parents.length - 1];
        if (node.parent) {
            var t = node.parent.type;
            if (typeof type == "string") {
                return type === t;
            } else {
                return type.indexOf(t) != -1;
            }
        }
        return false;
    }

    var errors = {};
    var errorCount = 0;
    var scopes = [];
    var scopeStack = [];

    // lowest rank in stack is the declaration if <= 3
    function addSymbol(name, range, declarationRank)
    {
        name += '_';
        var scopeIdx = scopeStack.length - 1;
        var found = false;
        while (true) {
            if (scopeStack[scopeIdx].objects[name] !== undefined) {
                found = true;
                break;
            } else if (declarationRank == 0 || scopeIdx == 0) {
                break;
            }
            --scopeIdx;
        }

        // console.log("Adding symbol", name, range, location(range), found, declarationRank, parents[parents.length - 2].type, parents[parents.length - 1].type);
        range.push(declarationRank);
        if (found) {
            scopeStack[scopeIdx].objects[name].push(range);
        } else {
            scopeStack[scopeIdx].objects[name] = [range];
        }
        ++scopeStack[scopeIdx].count;
    }

    function qualifiedName(node)
    {
        if (!node)
            node = parents[parents.length - 1];
        var orig = node;
        var seen = [];
        function resolveName(n)
        {
            if (seen.indexOf(n) != -1)
                return undefined;
            seen.push(n);
            if (n) {
                switch (n.type) {
                case esprima.Syntax.Identifier:
                    return n.name;
                case esprima.Syntax.Literal:
                    return n.value;
                case esprima.Syntax.MemberExpression:
                    if (seen.indexOf(n.object) != -1 || seen.indexOf(n.property) != -1)
                        break;
                    return resolveName(n.object) + "." + resolveName(n.property);
                case esprima.Syntax.ObjectExpression:
                case esprima.Syntax.Property:
                    return resolveName(n.key);
                case esprima.Syntax.VariableDeclarator:
                case esprima.Syntax.FunctionDeclaration:
                    return resolveName(n.id);
                case esprima.Syntax.AssignmentExpression:
                    return resolveName(n.left);
                case esprima.Syntax.CallExpression:
                    return resolveName(n.callee);
                default:
                    break;
                    // console.log("Not sure how to resolve", node.type, node.range);
                }
            }
            return undefined;
        }

        var name = undefined;
        var prev = undefined;
        while (node) {
            if (node.type) {
                var done = false;
                switch (node.type) {
                case esprima.Syntax.FunctionExpression:
                case esprima.Syntax.FunctionDeclaration:
                case esprima.Syntax.CallExpression:
                    done = prev;
                    break;
                case esprima.Syntax.AssignmentExpression:
                    done = prev && (prev.type != esprima.Syntax.ObjectExpression || node.right != prev);
                    // if (done)
                    //     console.log("Stopping", orig.range, "at", name, "because of", node.type);
                    break;
                case esprima.Syntax.VariableDeclarator:
                    done = prev && prev.type != esprima.Syntax.ObjectExpression && isChild("init", prev);
                    // if (done)
                    //     console.log("Stopping", orig.range, "at", name, "because of VariableDeclarator");
                    break;
                case esprima.Syntax.MemberExpression:
                    done = node.computed;
                    break;
                default:
                    break;
                }
                if (done)
                    break;

                var n = resolveName(node);
                if (n) {
                    if (!name) {
                        name = n;
                    } else {
                        name = n + "." + name;
                    }
                }
                prev = node;
                node = node.parent;
            }
        }
        return name;
    }

    function indexIdentifier(node)
    {
        node.indexed = true;
        var rank = 3;
        var name = undefined;
        switch (node.parent.type) {
        case esprima.Syntax.VariableDeclarator:
            if (isChild("init")) {
                rank = 5;
            } else {
                rank = 0;
            }
            break;
        case esprima.Syntax.FunctionDeclaration:
        case esprima.Syntax.FunctionExpression:
        case esprima.Syntax.Property:
            rank = 0;
            break;
        case esprima.Syntax.AssignmentExpression:
            rank = 3;
            break;
        case esprima.Syntax.CallExpression:
        case esprima.Syntax.UnaryExpression:
        case esprima.Syntax.BinaryExpression: // ### ???
        case esprima.Syntax.ReturnStatement:
        case esprima.Syntax.NewExpression:
        case esprima.Syntax.UpdateExpression:
        case esprima.Syntax.ForInStatement:
        case esprima.Syntax.IfStatement:
        case esprima.Syntax.WhileStatement:
        case esprima.Syntax.DoWhileStatement:
        case esprima.Syntax.ForStatement:
        case esprima.Syntax.LogicalExpression:
        case esprima.Syntax.ConditionalExpression:
        case esprima.Syntax.ArrayExpression:
            rank = 5;
            break;
        case esprima.Syntax.MemberExpression:
            if (isChild("property", node))
                name = qualifiedName(node.parent);
            rank = (node.parent.parent.type == esprima.Syntax.AssignmentExpression ? 3 : 5);
            break;
        default:
            // console.log("Shit", node.type);
            console.log("Unhandled parent", node.parent.type, node.parent.range, location(node.parent.range),
                        node.range, location(node.range));
            break;
        }
        if (name === undefined)
            name = qualifiedName();
        // console.log("Found identifier", name, node.range, parent.type, rank);
        addSymbol(name, node.range, rank);
    }

    estraverse.traverse(esrefactorContext._syntax, {
        enter: function (node) {
            if (parents.length)
                node.parent = parents[parents.length - 1];
            parents.push(node);
            var name = undefined;
            var declaration = 0;
            // console.log("entering", node.type, childKey(node));
            switch (node.type) {
            case esprima.Syntax.Program:
            case esprima.Syntax.FunctionExpression:
                node.scope = true;
                break;
            case esprima.Syntax.FunctionDeclaration:
                node.id.parent = node;
                indexIdentifier(node.id);
                node.scope = true;
                break;
            case esprima.Syntax.Identifier:
                if (!node.indexed) {
                    // we need to handle the function itself as part of the previous scope
                    indexIdentifier(node);
                }
                break;
            default:
                break;
            }
            if (node.scope) {
                var scope = { objects: {}, count: 0, type: node.type, index: scopes.length, defs: {} };
                scopeStack.push(scope);
                // console.log("adding a scope", node.type);
                scopes.push(scope);
            }
        },
        leave: function (node) {
            parents.pop();
            if (node.scope) {
                var scope = scopeStack.pop();
                scope.scopeStack = [];
                for (var i=0; i<scopeStack.length; ++i) {
                    scope.scopeStack.push(scopeStack[i].index);
                }
            }
        }
    });

    // var ret = { objects:[] };
    var ret = { symbols: [], symbolNames: {} };
    // console.log(scopes.length);
    // return ret;
    function add(name, locations, scope) {
        console.log("add", name, locations, scope.index, scope.scopeStack);
        if (!ret.symbolNames[name]) {
            ret.symbolNames[name] = locations;
        } else {
            ret.symbolNames[name] = ret.symbolNames[name].concat(locations);
        }

        var i;
        var defObj;
        var newDef = false;
        if (locations[0][2] < 5) {
            defObj = { location: locations[0], definition: true, name: name, references: [] };
            newDef = true;
        }
        if (locations[0][2] === 0) {
            scope.defs[name] = defObj;
        } else { // not a perfect hit, we need to search
            for (var idx=scope.scopeStack.length - 1; idx>=0; --idx) {
                var parentScope = scopes[scope.scopeStack[idx]];
                var def = parentScope.defs[name];
                if (def) {
                    if (def.location[2] === 0) {
                        newDef = false;
                        defObj = def;
                        break;
                    } else if (def.location[2] < 5 && (!defObj || def.location[2] < defObj.location[2])) {
                        newDef = false;
                        defObj = def;
                    }
                }
            }
        }
        if (newDef) {
            scope.defs[name] = defObj;
            ret.symbols.push(defObj);
            i = 1;
        }
        while (i < locations.length) {
            var loc = locations[i];
            defObj.references.push(loc);
            var obj = { location: loc, name: name };
            if (defObj)
                obj.target = defObj.location;
            ret.symbols.push(obj);
            ++i;
        }
    }
    for (var s=0; s<scopes.length; ++s) {
        var scope = scopes[s];
        // console.log("Adding things for scope", s, scope.count, scopes.length);
        if (scope.count) {
            for (var name in scope.objects) {
                scope.objects[name].sort(function(l, r) {
                    var ret = l[2] - r[2];
                    if (!ret)
                        ret = l[0] - r[0];
                    return ret;
                });
                add(name, scope.objects[name], scope);
            }
        }
    }
    ret.symbols.sort(function(l, r) {
        var ret = l.location[0] - r.location[0];
        if (!ret)
            ret = l.location[1] - r.location[1];
        return ret;
    });

    if (verbose >= 2) {
        estraverse.traverse(esrefactorContext._syntax, {
            enter: function(node) {
                delete node.parent;
                delete node.indexed;
            }
        });
        ret.ast = esrefactorContext._syntax;
    }
    if (errors)
        ret.errors = errors;
    return ret;
}

function findLocation(symbols, offset) {
    if (!symbols)
        return undefined;
    // console.log("CALLED HERE", symbols);
    var min = 0;
    var max = symbols.length - 1;
    var cur;
    function compare() {
        var currentElement = symbols[cur];
        if (offset < currentElement.location[0]) {
            return 1;
        } else if (offset >= currentElement.location[0] && offset < currentElement.location[1]) {
            return 0;
        }
        return -1;
    }

    while (min <= max) {
        cur = (min + max) >> 1;
        switch (compare()) {
        case -1:
            min = cur + 1;
            break;
        case 1:
            max = cur - 1;
            break;
        default:
            return symbols[cur];
        }
    }
    return undefined;
}

module.exports = {
    indexFile: indexFile,
    findLocation: findLocation
};
