// rjs/rjs.js
module.exports = {
    // errors
    ERROR_OK: 0,
    ERROR_READFAILURE: 1,
    ERROR_PARSE_FAILURE: 2,
    ERROR_MISSING_FILE: 3,
    ERROR_PROTOCOL_ERROR: 4,
    ERROR_FILE_ALREADY_INDEXED: 5,
    ERROR_INVALID_LOCATION: 6,
    ERROR_FILE_NOT_INDEXED: 7,
    ERROR_SYMBOL_NOT_FOUND: 8,
    ERROR_MORE_DATA: 9,
    ERROR_STATFAILURE: 10,
    ERROR_MISSING_SYMBOLNAME: 11,

    errorCodeToString: function(code) {
        var that = this;
        for (var i in that) {
            if (that[i] === code && i.lastIndexOf("ERROR_", 0) === 0) {
                return i;
            }
        }
        return undefined;
    },

    MESSAGE_COMPILE: 1,
    MESSAGE_FOLLOW_SYMBOL: 2,
    MESSAGE_FIND_REFERENCES: 3,
    MESSAGE_DUMP: 4,
    MESSAGE_CURSOR_INFO: 5,
    MESSAGE_FIND_SYMBOLS: 6,
    MESSAGE_LIST_SYMBOLS: 7,

    messageTypeToString: function(code) {
        var that = this;
        for (var i in that) {
            if (that[i] === code && i.lastIndexOf("MESSAGE_", 0) === 0) {
                return i;
            }
        }
        return undefined;
    },

    defaultPort: 5678
};
